data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/DefaultSqlPreparedQuery.java

```diff
@@ -180,6 +216,152 @@ public void attachPageable(Pageable pageable, boolean isSingleResult) {
         }
     }
 
+    /**
+     * A utility method for reversing the sort.
+     *
+     * @param sort The current sort
+     * @return reversed sort
+     */
+    private Sort reverseSort(Sort sort) {
+        if (!sort.isSorted()) {
+            return sort;
+        }
+        List<Order> orders = new ArrayList<>();
+        for (Order order : sort.getOrderBy()) {
+            orders.add(new Order(
+                order.getProperty(),
+                order.getDirection() == Direction.ASC ? Direction.DESC : Direction.ASC,
+                order.isIgnoreCase()
+            ));
+        }
+        return Sort.of(orders);
+    }
+
+    /**
+     * Add relevant query clauses and query bindings to use cursored pagination.
+     *
+     * @param cursor The supplied cursor
+     * @param sort The sorting that will be used in the query
+     * @return The additional query part
+     */
+    @NonNull
+    private String buildCursorPagination(@Nullable List<Object> cursor, @NonNull Sort sort) {
+        List<Sort.Order> orders = sort.getOrderBy();
+        cursorProperties = new ArrayList<>();
+        for (Order order: orders) {
+            cursorProperties.add(getPersistentEntity().getPropertyByName(order.getProperty()));
+        }
+        if (cursor == null) {
+            return "";
+        }
+        if (orders.size() != cursor.size()) {
+            throw new IllegalArgumentException("The cursor must match the sorting size");
+        }
+        if (orders.isEmpty()) {
+            throw new IllegalArgumentException("At least one sorting property must be supplied");
+        }
+
+        List<QueryParameterBinding> cursorBindings = new ArrayList<>();
+        cursorQueryBindings = new ArrayList<>();
+        for (int i = 0; i < cursor.size(); ++i) {
+            cursorBindings.add(new CursoredQueryParameterBinder(
+                "cursor_" + i, cursorProperties.get(i).getDataType(), cursor.get(i)
+            ));
+        }
+
+        StringBuilder builder = new StringBuilder(" ");
+        if (query.contains("WHERE")) {
```

Comment:
Maybe we can generate the WHERE part with brackets for the cursor cases so this check is minimal